src/Tmds.DBus/SignalMatchRule.cs

```diff
@@ -1,27 +1,147 @@
 // Copyright 2006 Alp Toker <alp@atoker.com>
 // Copyright 2016 Tom Deseyn <tom.deseyn@gmail.com>
+// Copyright 2021 David Lechner <david@lechnology.com>
 // This software is made available under the MIT License
 // See COPYING for details
 
+using System;
+using System.Collections.Generic;
 using System.Text;
+using Tmds.DBus.Protocol;
 
 namespace Tmds.DBus
 {
-    internal class SignalMatchRule
+    /// <summary>
+    /// Match rules describe the messages that should be sent to a client,
+    /// based on the contents of the message.
+    /// </summary>
+    public sealed class SignalMatchRule
     {
+        /// <summary>
+        /// Match messages sent over or to a particular interface.
+        /// </summary>
         public string Interface { get; set; }
+
+        /// <summary>
+        /// Match messages which have the give method or signal name.
+        /// </summary>
         public string Member { get; set; }
+
+        /// <summary>
+        /// Match messages which are sent from or to the given object.
+        /// </summary>
         public ObjectPath? Path { get; set; }
 
+        /// <summary>
+        /// Match messages which are sent from or to an object for which the 
+        /// object path is either the given value, or that value followed by one
+        /// or more path components. 
+        /// </summary>
+        public ObjectPath? PathNamespace { get; set; }
+
+        /// <summary>
+        /// Arg matches are special and are used for further restricting the
+        /// match based on the arguments in the body of a message.
+        /// </summary>
+        public  (int index, string arg)[] Args { get; set; }
+
+        /// <summary>
+        /// Argument path matches provide a specialized form of wildcard
+        /// matching for path-like namespaces.
+        /// </summary>
+        public (int index, string argPath)[] ArgPaths { get; set; }
+
+        /// <summary>
+        /// Match messages whose first argument is of type STRING, and is a bus
+        /// name or interface name within the specified namespace.
+        /// </summary>
+        public string Arg0Namespace { get; set; }
+
+        internal bool MatchHeader(Header header) 
+        {
+            if (header.MessageType != MessageType.Signal)
+            {
+                return false;
+            }
+
+            if (header.Interface != Interface)
+            {
+                return false;
+            }
+
+            if (header.Member != Member)
+            {
+                return false;
+            }
+
+            if (Path.HasValue && header.Path != Path)
+            {
+                return false;
+            }
+
+            if (PathNamespace.HasValue && (!header.Path.HasValue ||
+                !header.Path.Value.Value.StartsWith(PathNamespace.Value.Value, StringComparison.Ordinal)) )
```

Comment:
We need to do more than `StartsWith`:

From the spec:
>  For example, path_namespace='/com/example/foo' would match signals sent by /com/example/foo or by /com/example/foo/bar, but not by /com/example/foobar.

Reply:
Good catch. I suppose the same applies to `arg0namespace` with `.` instead of `/` as a delimiter.

Reply:
Yes, everywhere we're using `StartsWith`.